feat: snuba version of group events

alex-hofsteede opened the pull request:
This provides a snuba implementation of the group events api that should maybe be faster than the django based version, and allow some of our larger customers to use the group events tab. 

The queries it makes should benefit from the `split_query` optimization for reverse-time-sorted data that snuba does.

alex-hofsteede committed:
feat: Snuba version of group events

alex-hofsteede commented on src/sentry/api/endpoints/group_events.py:
```diff
@@ -115,4 +140,43 @@ def respond(queryset):
+            paginator_cls=DateTimePaginator,
+        )
+
+    def _get_environment(self, request, group):
```
This is maybe an unnecessary layer of indirection.

alex-hofsteede commented:
@mitsuhiko The snuba version of this isn't tested yet, but this should hopefully solve the timeout/latency issues on the issue events tab.

alex-hofsteede committed:
test: Tests for snuba GroupEvents endpoint

alex-hofsteede commented on src/sentry/api/endpoints/group_events.py:
```diff
@@ -38,63 +43,81 @@ def get(self, request, group):
-            )
-
-        events = Event.objects.filter(group_id=group.id)
+        backend = request.COOKIES.get('eventstream', 'legacy')
```
This is one thing that we would need to change before shipping, Either moving this to an option so that we can change it globally at runtime, or just hardcoding snuba as the default

alex-hofsteede committed:
use options instead of cookies

wedamija commented on src/sentry/api/endpoints/group_events.py:
```diff
@@ -38,63 +43,78 @@ def get(self, request, group):
-                query_kwargs = parse_query([group.project], raw_query, request.user)
-            except InvalidQuery as exc:
-                return Response({'detail': six.text_type(exc)}, status=400)
+            environment = self._get_environment(request, group)
```
Does it make sense to move these calls outside and to pass the variables in?

alex-hofsteede replied:
you mean move them up into the `get()`? yeah that would probably make sense and be a little DRYer

alex-hofsteede replied:
done

alex-hofsteede committed:
cleanup and getting tags for returned SnubaEvents

alex-hofsteede commented on src/sentry/api/endpoints/group_events.py:
```diff
@@ -38,63 +43,78 @@ def get(self, request, group):
+        return self.paginate(
+            request=request,
+            on_results=lambda results: serialize(
+                [SnubaEvent(row) for row in results], request.user),
```
I added a change to make `SnubaEvent` fetch all tags. This PR is supposed to provide a compatible implementation for the group events page so I should do it here. 

It will all be replaced by the aformentioned PR, but who knows when that will get reviewed/shipped

alex-hofsteede committed:
review feedback

alex-hofsteede committed:
fix tests

alex-hofsteede committed:
Only grab tags in this specific endpoint

Just in case it creates problems for other endpoints using
SnubaEvent.selected_columns

alex-hofsteede commented:
@wedamija it turns out the UI code needs a couple more things from the serialized SnubaEvent than I was providing, so this doesn't quite work yet. but feel free to put your environment changes on top of this now, the changes I have to make are all in the serializer, and shouldn't affect the `group_events.py` code.
